File: packages/dsa-ts/src/structures/MinHeap.ts
/**
 * Min Heap — árbol binario completo donde el padre ≤ hijos.
 * Base del PriorityQueue.
 *
 * Complexity:
 *   insert     O(log n)
 *   extractMin O(log n)
 *   peekMin    O(1)
 */
export class MinHeap<T> {
  private readonly heap: T[] = [];

  constructor(private readonly comparator: (a: T, b: T) => number) {}

  get size(): number    { return this.heap.length; }
  get isEmpty(): boolean { return this.heap.length === 0; }

  /** Devuelve el mínimo sin extraerlo. O(1) */
  peekMin(): T | undefined {
    return this.heap[0];
  }

  /** Inserta un elemento y reordena. O(log n) */
  insert(value: T): void {
    this.heap.push(value);
    this.bubbleUp(this.heap.length - 1);
  }

  /** Extrae y devuelve el mínimo. O(log n) */
  extractMin(): T | undefined {
    if (this.heap.length === 0) return undefined;
    const min = this.heap[0] as T;
    const last = this.heap.pop() as T;
    if (this.heap.length > 0) {
      this.heap[0] = last;
      this.sinkDown(0);
    }
    return min;
  }

  private bubbleUp(index: number): void {
    while (index > 0) {
      const parent = Math.floor((index - 1) / 2);
      if (this.comparator(this.heap[index] as T, this.heap[parent] as T) < 0) {
        this.swap(index, parent);
        index = parent;
      } else {
        break;
      }
    }
  }

  private sinkDown(index: number): void {
    const length = this.heap.length;
    while (true) {
      const left  = 2 * index + 1;
      const right = 2 * index + 2;
      let smallest = index;

      if (left < length &&
        this.comparator(this.heap[left] as T, this.heap[smallest] as T) < 0) {
        smallest = left;
      }
      if (right < length &&
        this.comparator(this.heap[right] as T, this.heap[smallest] as T) < 0) {
        smallest = right;
      }
      if (smallest === index) break;
      this.swap(index, smallest);
      index = smallest;
    }
  }

  private swap(i: number, j: number): void {
    const tmp     = this.heap[i] as T;
    this.heap[i]  = this.heap[j] as T;
    this.heap[j]  = tmp;
  }

  toArray(): T[] {
    return [...this.heap];
  }
}

File: packages/dsa-py/src/structures/min_heap.py
"""
Min Heap — árbol binario completo donde padre ≤ hijos.

Complexity:
  insert      O(log n)
  extract_min O(log n)
  peek_min    O(1)
"""
from typing import Generic, TypeVar, Callable

T = TypeVar("T")


class MinHeap(Generic[T]):
    def __init__(self, comparator: Callable[[T, T], int] | None = None) -> None:
        self._heap: list[T] = []
        self._cmp  = comparator or (lambda a, b: -1 if a < b else (1 if a > b else 0))  # type: ignore[operator]

    @property
    def size(self) -> int:
        return len(self._heap)

    @property
    def is_empty(self) -> bool:
        return len(self._heap) == 0

    def peek_min(self) -> T | None:
        return self._heap[0] if self._heap else None

    def insert(self, value: T) -> None:
        self._heap.append(value)
        self._bubble_up(len(self._heap) - 1)

    def extract_min(self) -> T | None:
        if not self._heap:
            return None
        minimum = self._heap[0]
        last    = self._heap.pop()
        if self._heap:
            self._heap[0] = last
            self._sink_down(0)
        return minimum

    def _bubble_up(self, i: int) -> None:
        while i > 0:
            parent = (i - 1) // 2
            if self._cmp(self._heap[i], self._heap[parent]) < 0:
                self._heap[i], self._heap[parent] = self._heap[parent], self._heap[i]
                i = parent
            else:
                break

    def _sink_down(self, i: int) -> None:
        n = len(self._heap)
        while True:
            smallest = i
            for child in (2 * i + 1, 2 * i + 2):
                if child < n and self._cmp(self._heap[child], self._heap[smallest]) < 0:
                    smallest = child
            if smallest == i:
                break
            self._heap[i], self._heap[smallest] = self._heap[smallest], self._heap[i]
            i = smallest
